### PR TITLE
Optimize some functions in std::process

### PR DESCRIPTION
**Pull request by alexcrichton**
- Be sure that `read_to_end` gets directed towards `read_to_end_uninitialized` for all handles
- When spawning a child that guaranteed doesn't need a stdin, don't actually create a stdin pipe for that process, instead just redirect it to /dev/null
- When calling `wait_with_output`, don't spawn threads to read out the pipes of the child. Instead drain all pipes on the calling thread and _then_ wait on the process.

Functionally, it is intended that nothing changes as part of this PR

---

Note that this was the same as #31613, and even after that it turned out that fixing Windows was easier than I thought! To copy a comment from over there:

> As some rationale for this as well, it's always bothered me that we've spawned threads in the standard library for this (seems a bit overkill), and I've also been curious lately as to our why our build times for Windows are so much higher than Unix (on the buildbots we have). I have done basically 0 investigation into why, but I figured it can't help to try to optimize Command::output which I believe is called quite a few times during the test suite.

### CODE REVIEW

**Comment by alexcrichton** [comment]
@bors: r=brson 89f3f0e

I think I basically have no idea how overlapped I/O works.

**Comment by alexcrichton** [comment]
@bors: r- 

ah it's probably legit, this patch definitely deals with functions that would cause the test suite to deadlock there

File: src/libstd/sys/windows/pipe.rs [hunk 1]
```diff
@@ -41,7 +110,193 @@ impl AnonPipe {
...
+        // To do that, we call `CancelIo` to cancel any pending operation, and
+        // if that succeeds we wait for the overlapped result.
+        //
+        // If anything here fails, there's not really much we can do, so we leak
+        // the buffer/OVERLAPPED pointers to ensure we're at least memory safe.
+        if self.pipe.cancel_io().is_err() || self.result().is_err() {
+            let buf = mem::replace(self.dst, Vec::new());
+            let overlapped = Box::new(unsafe { mem::zeroed() });
+            let overlapped = mem::replace(&mut self.overlapped, overlapped);
+            mem::forget((buf, overlapped));
```
Comment (brson) [inline src/libstd/sys/windows/pipe.rs]: This is so scary. Presumably this will never, ever run, until someday it does.

**Comment by alexcrichton** [comment]
@bors: r=brson 96f181118140e251ce5aa1f9e55c95c76ed0f1e7

Now with more assertions!

**Comment by alexcrichton** [comment]
@bors: r=brson 6b35d31

Life lesson learned! When `ReadFile` returns with `ERROR_BROKEN_PIPE` the event in the `OVERLAPPED` is apparently set to "not signaled". This would cause `WaitForMultipleObjects` to never again say that object was finished, so we'd never recognize the EOF in the `result()` function.

This switches up the logic so if `result()` or `schedule_read()` ever returns EOF we consider ourselves immediately done with that pipe and never hit `WaitForMultipleObjects` again.

**Comment by alexcrichton** [comment]
@bors: r=brson 7c3038f

whoops typos...

**Comment by alexcrichton** [comment]
:astonished:

**Pull request merged**